pipeline/subtraction.py

```diff
@@ -36,6 +44,79 @@ def __init__(self, **kwargs):
 
         # TODO: add a reference cache here.
 
+    @staticmethod
+    def _subtract_naive(new_image, ref_image):
+        """Subtract the reference from the image directly, assuming they are aligned and same shape.
+
+        Doesn't do any fancy PSF matching or anything, just takes the difference of the data arrays.
+
+        Parameters
+        ----------
+        new_image : Image
+            The Image containing the new data, including the data array, weight, and flags
+        ref_image : Image
+            The Image containing the reference data, including the data array, weight, and flags
+
+        Returns
+        -------
+        outim : np.ndarray
+            The difference between the new and reference images
+        outwt: np.ndarray
+            The weight image for the difference
+        outfl: np.ndarray
+            The flag image for the difference
+        """
+        outim = new_image.data - ref_image.data
+
+        # must add the variance to make a new weight image
+        new_mask = new_image.weight <= 0
+        new_weight = new_image.weight.copy()
+        new_weight[new_mask] = np.nan
+        new_var = 1 / new_weight ** 2
+
+        ref_mask = ref_image.weight <= 0
+        ref_weight = ref_image.weight.copy()
+        ref_weight[ref_mask] = np.nan
+        ref_var = 1 / ref_weight ** 2
+
+        outwt = 1 / np.sqrt(new_var + ref_var)
+        outwt[new_mask] = 0  # make sure to make zero weight the pixels that started out at zero weight
+
+        outfl = new_image.flags.copy()
+        outfl |= ref_image.flags
+
+        return outim, outwt, outfl
+
+    def _subtract_zogy(self, new_image, ref_image):
+        """Use ZOGY to subtract the two images.
+
+        This applies PSF matching and uses the ZOGY algorithm to subtract the two images.
+        reference: https://ui.adsabs.harvard.edu/abs/2016ApJ...830...27Z/abstract
+
+        Parameters
+        ----------
+        new_image : Image
+            The Image containing the new data, including the data array, weight, and flags.
+            Image must also have the PSF and ZeroPoint objects loaded.
+        ref_image : Image
+            The Image containing the reference data, including the data array, weight, and flags
+            Image must also have the PSF and ZeroPoint objects loaded.
+
+        Returns
+        -------
+        outim : np.ndarray
+            The difference between the new and reference images
+        outwt: np.ndarray
+            The weight image for the difference
+        outfl: np.ndarray
+            The flag image for the difference
+        score: np.ndarray
+            The ZOGY score image (the matched-filter result)
+        zogy_psf: np.ndarray
+            The ZOGY PSF image (the matched-filter PSF)
+        """
+        pass
```

Comment:
I thought I remembered seeing a ZOGY implementation in an earlier PR; where was that?  Why not use it here?

Reply:
(But, in any event, probably a good idea to throw a NotImplementedError rather than just pass.)

Reply:
Ahh yes, I was half way into this PR when we started working on the backlog. I will put a not-implemented here and we will have to continue this in a new PR that is less about the data flow and more about the actual subtraction.